src/pymor/algorithms/qr.py

```diff
@@ -0,0 +1,67 @@
+# This file is part of the pyMOR project (https://www.pymor.org).
+# Copyright pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
+
+from pymor.algorithms.gram_schmidt import gram_schmidt
+
+
+def qr(A, product=None, offset=0, check=True, check_tol=1e-3, copy=True):
+    """Compute a QR decomposition.
+
+    Parameters
+    ----------
+    A
+        The |VectorArray| for which to compute the QR decomposition.
+    product
+        The inner product |Operator| w.r.t. which to orthonormalize.
+        If `None`, the Euclidean product is used.
+    offset
+        Assume that the first `offset` vectors are already orthonormal and start the
+        algorithm at the `offset + 1`-th vector.
+    check
+        If `True`, check if the resulting |VectorArray| is really orthonormal.
```

Comment:
Should we also check if $QR \approx A$? I'd say 'yes' in some way. I.e., could be enabled/disabled by a separate `check` argument or not.

Reply:
Sounds reasonable to me. I'll think more about the types of `check` arguments.

Reply:
> Sounds reasonable to me. I'll think more about the types of `check` arguments.

One thing to also think about could be to reuse the orthogonality error, if the algorithm computes that. E.g. in Cholesky QR, it is computed in the adaptive version.
Another thing is that nonzero `offset` can make the error computations cheaper.